sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/SqlTypes.java

```diff
@@ -27,4 +28,7 @@ private SqlTypes() {}
 
   /** Beam LogicalType corresponding to ZetaSQL/CalciteSQL DATE type. */
   public static final LogicalType<LocalDate, Long> DATE = new Date();
+
+  /** Beam LogicalType corresponding to ZetaSQL/CalciteSQL TIME type. */
+  public static final LogicalType<LocalTime, Long> TIME = new Time();
```

Comment:
This is the most important line of this CL for correctness, wanted to get attention on it early.

ZetaSQL uses a very specific format for this, which roughly matches the struct in LocalTime: https://github.com/google/zetasql/blob/79adcd0fe227173e68ed7aa88f580a691ebe82c2/zetasql/public/civil_time.h#L85

It appears the Long base type is an offset in Nanoseconds. Is that conversion lossless? Is it expensive?

Reply:
Now the internal build passes.

Thanks for the question. Conversion is lossless. Actually it is done by calling ZetaSQL code: `CivilTimeEncoder.decodePacked64TimeNanosAsJavaTime()` and `encodePacked64TimeNanos()`. See conversion code in `ZetaSqlBeamTranslationUtils.java`. I looked at the code and it seems to be purely bit manipulation, so I guess it's not very expensive.

Reply:
Great! Bitshifting for `java.time.LocalTime` is lossless and fast. Looks like the conversion to/from Long is done by `LocalTime.ofNanoOfDay` and `LocalTime.toNanoOfDay` and those do math with a few constants that is also lossless and fast.

Reply:
It looks like one or more of the conversions (at very least `LocalTime` to `Long`) is not lossless, it corrupts leap seconds. Here is a set of trivial identity test cases that needs to pass any type conversion:

```
SELECT
    CAST(CAST(TIME "00:00:00" as STRING) as TIME),
    CAST(CAST(TIME "23:59:59.999999" as STRING) as TIME),
    CAST(CAST(TIME "01:02:03.123456" as STRING) as TIME),
    CAST(CAST(TIME "01:02:03.12345" as STRING) as TIME),
    CAST(CAST(TIME "01:02:03.120000" as STRING) as TIME),
    CAST(CAST(TIME "01:02:3" as STRING) as TIME),
    CAST(CAST(TIME "01:2:03" as STRING) as TIME),
    CAST(CAST(TIME "1:02:03" as STRING) as TIME),
    #Leap second cases
    CAST(CAST(TIME "23:59:60" as STRING) as TIME),
    CAST(CAST(TIME "12:59:60" as STRING) as TIME),
    CAST(CAST(TIME "12:59:60.123456" as STRING) as TIME),

    CAST(CAST("00:00:00" as TIME) as STRING),
    CAST(CAST("23:59:59.999999" as TIME) as STRING),
    CAST(CAST("01:02:03.123456" as TIME) as STRING),
    CAST(CAST("01:02:03.12345" as TIME) as STRING),
    CAST(CAST("01:02:03.120000" as TIME) as STRING),
    CAST(CAST("01:02:3" as TIME) as STRING),
    CAST(CAST("01:2:03" as TIME) as STRING),
    CAST(CAST("1:02:03" as TIME) as STRING),
    #Leap second cases
    CAST(CAST("23:59:60" as TIME) as STRING),
    CAST(CAST("12:59:60" as TIME) as STRING),
    CAST(CAST("12:59:60.123456" as TIME) as STRING)
```

Reply:
Actually, the ZetaSQL standard calls for those lossy conversions. The test is failing due to truncation of precision to millisecond. Expected right answers for this test case:
```
  {
    "f0_": "00:00:00",
    "f1_": "23:59:59.999999",
    "f2_": "01:02:03.123456",
    "f3_": "01:02:03.123450",
    "f4_": "01:02:03.120000",
    "f5_": "01:02:03",
    "f6_": "01:02:03",
    "f7_": "01:02:03",
    "f8_": "00:00:00",
    "f9_": "13:00:00",
    "f10_": "13:00:00",
    "f11_": "00:00:00",
    "f12_": "23:59:59.999999",
    "f13_": "01:02:03.123456",
    "f14_": "01:02:03.123450",
    "f15_": "01:02:03.120",
    "f16_": "01:02:03",
    "f17_": "01:02:03",
    "f18_": "01:02:03",
    "f19_": "00:00:00",
    "f20_": "13:00:00",
    "f21_": "13:00:00"
  }
```

Reply:
Thanks for the pointer. This test is passing now!